pkg/function/client/client.go

```diff
@@ -79,20 +81,51 @@ func (c *client) MapTFn(ctx context.Context, datum *functionpb.Datum) ([]*functi
 }
 
 // ReduceFn applies a reduce function to a datum stream.
-func (c *client) ReduceFn(ctx context.Context, datumStreamCh <-chan *functionpb.Datum) ([]*functionpb.Datum, error) {
+func (c *client) ReduceFn(ctx context.Context, datumStreamCh <-chan *functionpb.Datum) (*functionpb.DatumList, error) {
+	var g errgroup.Group
+	datumList := &functionpb.DatumList{}
+
 	stream, err := c.grpcClt.ReduceFn(ctx)
 	if err != nil {
 		return nil, fmt.Errorf("failed to execute c.grpcClt.ReduceFn(): %w", err)
 	}
-	for datum := range datumStreamCh {
-		if err := stream.Send(datum); err != nil {
-			return nil, fmt.Errorf("failed to execute stream.Send(%v): %w", datum, err)
+	g.Go(func() error {
+		var sendErr error
+		for datum := range datumStreamCh {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+				if sendErr = stream.Send(datum); sendErr != nil {
+					return sendErr
```

Comment:
Does it need to `CloseSend()`?

Reply:
No we don't need to. 
I looked at their implementation they call finish on the stream if there are any errors
```
	defer func() {
		if err != nil && err != io.EOF {
			// Call finish on the client stream for errors generated by this SendMsg
			// call, as these indicate problems created by this client.  (Transport
			// errors are converted to an io.EOF error in csAttempt.sendMsg; the real
			// error will be returned from RecvMsg eventually in that case, or be
			// retried.)
			cs.finish(err)
		}
	}()
```

Reply:
could you please note that in the comment?